File: python/src/hdlinfo/protocols/amba/axi4.py
from ...json import *
from dataclasses import field
from typing import List
from functools import cached_property
from ._common import is_pow2

__all__ = ["Config", "kind"]

kind = "axi4"


_AXI4_SIGNALS = [
    "AWVALID",
    "AWREADY",
    "AWADDR",
    "AWPROT",
    "AWUSER",
    "AWREGION",
    "AWQOS",
    "AWCACHE",
    "AWBURST",
    "AWSIZE",
    "AWLEN",
    "AWID",
    "AWLOCK",
    "WVALID",
    "WREADY",
    "WDATA",
    "WSTRB",
    "WUSER",
    "WLAST",
    "BVALID",
    "BREADY",
    "BRESP",
    "BUSER",
    "BID",
    "ARVALID",
    "ARREADY",
    "ARADDR",
    "ARPROT",
    "ARUSER",
    "ARREGION",
    "ARQOS",
    "ARCACHE",
    "ARBURST",
    "ARSIZE",
    "ARLEN",
    "ARID",
    "ARLOCK",
    "RVALID",
    "RREADY",
    "RDATA",
    "RRESP",
    "RUSER",
    "RID",
    "RLAST"
]

_AXI4LITE_SIGNALS = [
    "AWVALID",
    "AWREADY",
    "AWADDR",
    "AWPROT",
    "WVALID",
    "WREADY",
    "WDATA",
    "WSTRB",
    "BVALID",
    "BREADY",
    "BRESP",
    "ARVALID",
    "ARREADY",
    "ARADDR",
    "ARPROT",
    "RVALID",
    "RREADY",
    "RDATA",
    "RRESP"
]


@register_dataclass_adv("hdlinfo.protocols.amba.axi4.Config", aliases=["chext.amba.axi4.Config"])
@dataclass(frozen=True)
class Config:
    wId: int = 0
    wAddr: int = 32
    wData: int = 32
    read: bool = True
    write: bool = True
    lite: bool = False
    wUserAR: int = 0
    wUserR: int = 0
    wUserAW: int = 0
    wUserW: int = 0
    wUserB: int = 0

    skippedSignals: List[str] = field(default_factory=list)

    def __post_init__(self):
        if self.wData < 8:
            raise ValueError("wData must be at least 8")
        if not is_pow2(self.wData):
            raise ValueError("wData must be a power of 2")
        if self.lite and self.wData not in [32, 64]:
            raise ValueError("If lite is True, wData must be 32 or 64")
        if self.wAddr <= 0:
            raise ValueError("wAddr must be greater than 0")
        if any(user < 0 for user in [self.wUserAR, self.wUserR, self.wUserAW, self.wUserW, self.wUserB]):
            raise ValueError("all wUserX attributes must be greater than or equal to 0")

    @property
    def wStrobe(self):
        return self.wData // 8

    @property
    def signals(self) -> List[str]:
        if self.lite:
            return list(_AXI4LITE_SIGNALS)

        result = []
        for signal in _AXI4_SIGNALS:
            if signal in self.skippedSignals:
                continue

            if signal.endswith("USER"):
                width: int = getattr(self, f'wUser{signal.removesuffix("USER")}')
                if width == 0:
                    continue
            else:
                result.append(signal)
        return result

File: scala/src/main/scala/hdlinfo/protocols/amba/axi4.scala
package hdlinfo.protocols.amba.axi4

import hdlinfo.Registry
import hdlinfo.util.isPow2

/** AXI4 interface configuration.
  *
  * @param wId
  *   identification tag width
  * @param wAddr
  *   address width
  * @param wData
  *   data bus width
  * @param read
  *   enable read channels
  * @param write
  *   enable write channels
  * @param lite
  *   use AXI4-Lite variant
  * @param hasLock
  *   enable ARLOCK and AWLOCK
  * @param hasCache
  *   enable ARCACHE and AWCACHE
  * @param hasProt
  *   enable ARPROT and AWPROT
  * @param hasQos
  *   enable ARQOS and AWQOS
  * @param hasRegion
  *   enable ARREGION and AWREGION
  * @param axi3Compat
  *   AXI3 compatibility: 2-bit AxLOCK and 4-bit AxLEN, still without WID
  * @param wUserAR
  *   ARUSER field width
  * @param wUserR
  *   RUSER field width
  * @param wUserAW
  *   AWUSER field width
  * @param wUserW
  *   WUSER field width
  * @param wUserB
  *   BUSER field width
  */
case class Config(
    val wId: Int = 0,
    val wAddr: Int = 32,
    val wData: Int = 32,
    val read: Boolean = true,
    val write: Boolean = true,
    val lite: Boolean = false,
    val hasLock: Boolean = true,
    val hasCache: Boolean = true,
    val hasProt: Boolean = true,
    val hasQos: Boolean = true,
    val hasRegion: Boolean = true,
    val axi3Compat: Boolean = false,
    val wUserAR: Int = 0,
    val wUserR: Int = 0,
    val wUserAW: Int = 0,
    val wUserW: Int = 0,
    val wUserB: Int = 0
) {
  require(wData >= 8)
  require(isPow2(wData))
  require(!lite || (wData == 32 || wData == 64))

  /** width of the strobe signal for the write data channel */
  val wStrobe = wData / 8

  private def maybeZero(p: Boolean, w: Int) = if (p) w else 0

  val wLen = if (axi3Compat) 4 else 8
  val wLock = maybeZero(hasLock, if (axi3Compat) 2 else 1)
  val wCache = maybeZero(hasCache, 4)
  val wProt = maybeZero(hasProt, 3)
  val wQos = maybeZero(hasQos, 4)
  val wRegion = maybeZero(hasRegion, 4)
}

import io.circe.syntax._
import io.circe.generic.auto._

object register {
  def apply(): Unit = {
    Registry.register[Config]("hdlinfo.protocols.amba.axi4.Config")
    Registry.registerStringToType[Config]("chext.amba.axi4.Config")
  }
}
